go/services/reports/pkg/openai/batch_poller.go

```diff
@@ -0,0 +1,219 @@
+	"fmt"
+	"github.com/Magpie-Monitor/magpie-monitor/pkg/envs"
+	"go.uber.org/zap"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+)
+
+const (
+	OpenAiBatchStatus__Validating = "validating"
+	OpenAiBatchStatus__Failed     = "failed"
+	OpenAiBatchStatus__InProgress = "in_progress"
+	OpenAiBatchStatus__Finalizing = "finalizing"
+	OpenAiBatchStatus__Completed  = "completed"
+	OpenAiBatchStatus__Expired    = "expired"
+	OpenAiBatchStatus__Cancelling = "cancelling"
+	OpenAiBatchStatus__Cancelled  = "cancelled"
+)
+
+const BATCH_AWAITING_INTERVAL_SECONDS = 10
+
+type BatchPoller struct {
+	batches                chan *Batch
+	client                 *Client
+	pendingBatchRepository PendingBatchRepository
+	pollingIntervalSeconds int
+}
+
+func NewBatchPoller(client *Client, pendingBatchRepository PendingBatchRepository) *BatchPoller {
+
+	envs.ValidateEnvs("Missing envs for openai batch poller", []string{POLLING_INTERVAL_SECONDS_KEY})
+
+	pollingIntervalSeconds := os.Getenv(POLLING_INTERVAL_SECONDS_KEY)
+	pollingIntervalSecondsInt, err := strconv.Atoi(pollingIntervalSeconds)
+
+	if err != nil {
+		panic(fmt.Sprintf("%s is not a number", POLLING_INTERVAL_SECONDS_KEY))
+	}
+
+	poller := &BatchPoller{
+		batches:                make(chan *Batch),
+		client:                 client,
+		pendingBatchRepository: pendingBatchRepository,
+		pollingIntervalSeconds: pollingIntervalSecondsInt,
+	}
+
+	go poller.Start()
+
+	return poller
+}
+
+func (p *BatchPoller) Start() {
+	for {
+		batchIds, err := p.pendingBatchRepository.GetAllPending()
+		p.client.logger.Debug("Got batchIds from repository", zap.Any("ids", batchIds))
+
+		if err != nil {
+			p.client.logger.Error("Failed to get pending batches", zap.Error(err))
+			continue
+		}
+
+		for _, batchId := range batchIds {
+			batch, err := p.client.Batch(batchId)
+
+			if err != nil {
+				p.client.logger.Error("Failed to getch batch from OpenAI", zap.Error(err), zap.Any("batch", batch))
+				continue
+			}
+
+			if batch.isCompleted() {
+				p.client.logger.Debug("Batch has been completed", zap.Any("batch", batch))
+				p.pendingBatchRepository.CompleteBatch(batchId)
+			}
+
+			if batch.isFailed() {
+				p.client.logger.Error("Batch has been failed", zap.Any("batch", batch))
+				p.pendingBatchRepository.FailBatch(batchId)
+			}
+
+			if batch.isExpired() {
+				p.client.logger.Error("Batch has expired", zap.Any("batch", batch))
+				p.pendingBatchRepository.FailBatch(batchId)
+			}
+		}
+
+		p.client.logger.Info("Currenly pending batches", zap.Any("batchIds", batchIds))
+		time.Sleep(time.Second * time.Duration(p.pollingIntervalSeconds))
+	}
+}
+
+func (p *BatchPoller) Batch(batchId string) (*Batch, error) {
+
+	batch, err := p.pendingBatchRepository.GetPendingBatch(batchId)
+	if err != nil {
+		p.client.logger.Error("Failed to check if batch is pending", zap.String("batchId", batchId), zap.Error(err))
+		return nil, err
+	}
+
+	// If the batch is not in the repository then a nil will be returned
+	if batch != nil {
```

Comment:
If we "batch" a batchId that doesn't exist, shouldn't we return an error with a message? Easier to handle than nil, which can cause nilptr.

Reply:
I agree with this in most cases, but in pendingBatchRepository nil is a correct value of missing entry. This repository stores only pending batches, if one was finished then it will be missing (and it's not an error). I consider this nil as a None of a Option-like type. This "nil" is catched here and made sure that rest of the codebase never encounters a nil batch. An error will be something that actually needs handling (this is just a normal flow)

Reply:
Got it, thanks for explaining